workflows/datalogger/machine_specific/create_dozer_blade_poi_report_with_state/create_dozer_blade_poi_report_with_state.py

```diff
@@ -12,160 +32,70 @@ def path_up_to_last(a_last, a_inclusive=True, a_path=os.path.dirname(os.path.rea
 
-    for i, val in enumerate(assets[ac_uuid]["signatures"]):
-        if val["asset_urn"].startswith("urn:X-topcon:machine"):
-            machine_name = val["asset_urn"].split(":")[-1]
-
-        if val["asset_urn"].startswith("urn:X-topcon:device"):
-            device_id = val["asset_urn"].split(":")[-1]
-
-    utc_time = datetime.datetime.fromtimestamp(a_replicate["at"]/1000,tz=tz.gettz('Australia/Brisbane'))
-
-    position_quality = "Unknown"
-    if a_position_quality == 0:
-        position_quality = "Unknown"
-    elif a_position_quality == 1:
-        position_quality = "GPS Float"
-    elif a_position_quality == 2:
-        position_quality = "RTK Fixed"
-    elif a_position_quality == 3:
-        position_quality = "mm Enhanced"
-
-    a_file_ptr.write("\n-, {}, {}, {}, {}, {}, {}, {}, -, {}, {}, {}, {}, {}, {}, {}, {}, {}, {}".format(device_id, machine_name, utc_time, position_quality, a_position_error_horz, a_position_error_vert, a_auto_grade_control, a_reverse, delay_id, operator_id, task_id, a_blade_left[1], a_blade_left[0], a_blade_left[2], a_blade_right[1], a_blade_right[0], a_blade_right[2]))
-
+point_of_interest_dict = {}
 
+# Main datalogger data processing loop
+header_list = []
+line_count = 0
 for line in response.iter_lines():
-    decoded_line = base64.b64decode(line).decode('UTF-8')
-    logging.debug(decoded_line)
-    decoded_json = json.loads(decoded_line)
+    line_count += 1
+    decoded_json = json.loads(base64.b64decode(line).decode('UTF-8'))
 
     if decoded_json['type'] == "sitelink::State":
-        logging.debug("Found state.")
+        UpdateStateForAssetContext(a_state_msg=decoded_json, a_state_dict=state)
+        logging.debug("Found state. Current state: {}".format(json.dumps(state, indent=4)))
+
+    if decoded_json['type'] == "mfk::Replicate":
+        ProcessReplicate(a_decoded_json=decoded_json, a_resource_config_dict=resource_definitions, a_assets_dict=assets, a_state_dict=state, a_resources_dir=resources_dir, a_report_file=report_file, a_header_list=header_list, a_server=server, a_site_id=args.site_id, a_headers=headers, a_machine_description_filter="Generic Bulldozer (3DMC)")
 
-        if not decoded_json['data']['ac_uuid'] in state:
-            state[decoded_json['data']['ac_uuid']] = {}
+logging.info("Writing report to {}".format(args.datalogger_output_file_name))
 
-        if not decoded_json['data']['ns'] in state[decoded_json['data']['ac_uuid']]:
-            state[decoded_json['data']['ac_uuid']][decoded_json['data']['ns']] = {}
+report_file.close()
 
-        nested_state = {
-            "state" : decoded_json['data']['state'],
-            "value" : decoded_json['data']['value']
-        }
-        state[decoded_json['data']['ac_uuid']][decoded_json['data']['ns']][decoded_json['data']['state']] = nested_state
+with open(report_file_name, 'r+') as report_file:
+    content = report_file.read()
```

Comment:
this should open a new write file and loop line for line read old write new, for scalability - not nice reading potentially large files directly into memory.  if you're going to read the whole file into memory, you might as well stream directly into a list and not have to do this seek re-write bit at all

Reply:
Agreed, this is an inefficient implementation. Updated to iterate line by line after header write.